nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -761,53 +715,41 @@ static void
 ngx_http_modsecurity_detection_thread_func(void *data, ngx_log_t *log)
 {
     // Executed in a separate thread
-    ngx_http_modsecurity_detection_thread_ctx_t *thread_ctx = data;
+    ngx_http_modsecurity_thread_ctx_t *thread_ctx = data;
+    ngx_http_modsecurity_ctx_t *mod_ctx = ngx_http_get_module_ctx(thread_ctx->r, ngx_http_modsecurity);
 
     // Processing request headers
-    ngx_int_t rc = ngx_http_modsecurity_status(modsecProcessRequestHeaders(thread_ctx->req));
+    ngx_int_t rc = ngx_http_modsecurity_status(modsecProcessRequestHeaders(mod_ctx->req));
     if (rc != NGX_DECLINED) {
         return;
     }
 
-    if (modsecContextState(thread_ctx->req) == MODSEC_DISABLED) {
+    if (modsecContextState(mod_ctx->req) == MODSEC_DISABLED) {
         return;
     }
 
     // The name of modsecProcessRequestBody is a bit misleading. This function call is needed even to just process GET args.
-    modsecProcessRequestBody(thread_ctx->req);
+    modsecProcessRequestBody(mod_ctx->req);
 }
 
-
 static void
 ngx_http_modsecurity_detection_thread_completion(ngx_event_t *ev)
 {
     // executed in nginx event loop after thread task is done, in order to pick up and continue processing request
-    ngx_http_modsecurity_detection_thread_ctx_t *ctx = ev->data;
-
-    request_rec *req = ctx->req;
-    conn_rec *connection = ctx->connection;
-
-    if (req != NULL) {
-        modsecFinishRequest(req);
-    }
-    if (connection != NULL) {
-        modsecFinishConnection(connection);
-    }
+    ngx_http_request_t *r = ev->data;
+    --r->main->blocked;
+    ngx_http_finalize_request(r, NGX_DONE);
```

Comment:
We still call `ngx_http_core_run_phases(r)` in the prevention mode completion handler. This is needed to pass the request through the Nginx chain of handlers again once its offloaded processing is completed.

For detection mode, however, we don't need to do this because we return `NGX_DECLINED` from the handler immediately so the request keeps being handled and that is how we get it sent to the backend immediately without awaiting for the background task to complete.

`ngx_http_finalize_request()` is the documentation-recommended way to decrement the `count` and make sure the request is properly released when the ref count hits zero: https://nginx.org/en/docs/dev/development_guide.html#http_request

`r->aio` is not well documented, however, from what I could gather it effectiely tells Nginx there is some ongoing async I/O operation on this request so it does not fulfill the request. For prevention mode that was fine to do, though not sure if that was necessary but wouldn't hurt anyway. Here, it would cause the request to not be sent to the backend which is not exactly what we want.